File: ios-native/Jottery/Jottery/Models/NoteVersionRecord.swift
import Foundation
import GRDB

/// On-disk version snapshot. Maps directly to the `note_versions` SQLite table.
struct NoteVersionRecord: Codable, FetchableRecord, PersistableRecord, Identifiable {
    static let databaseTableName = "note_versions"

    // MARK: - Columns

    var versionKey: String         // "{noteId}:{version}" — PRIMARY KEY
    var noteId: String
    var version: Int
    var createdAt: String          // ISO 8601
    var syncedAt: String           // ISO 8601
    var content: String            // Encrypted JSON
    var tags: String               // Encrypted JSON (storage format — single blob)
    var attachments: String        // JSON array
    var syntaxLanguage: String?
    var wordWrap: Bool?
    var showPreview: Bool?
    var color: String?
    var reason: String             // "sync" | "manual-sync"
    var contentHash: String?
    var parentHash: String?

    var id: String { versionKey }

    // MARK: - Column Mapping

    enum Columns: String, ColumnExpression {
        case versionKey = "version_key"
        case noteId = "note_id"
        case version
        case createdAt = "created_at"
        case syncedAt = "synced_at"
        case content
        case tags
        case attachments
        case syntaxLanguage = "syntax_language"
        case wordWrap = "word_wrap"
        case showPreview = "show_preview"
        case color
        case reason
        case contentHash = "content_hash"
        case parentHash = "parent_hash"
    }

    // MARK: - Coding Keys

    enum CodingKeys: String, CodingKey {
        case versionKey = "version_key"
        case noteId = "note_id"
        case version
        case createdAt = "created_at"
        case syncedAt = "synced_at"
        case content
        case tags
        case attachments
        case syntaxLanguage = "syntax_language"
        case wordWrap = "word_wrap"
        case showPreview = "show_preview"
        case color
        case reason
        case contentHash = "content_hash"
        case parentHash = "parent_hash"
    }

    // MARK: - Factory

    /// Create a version snapshot from a NoteRecord.
    static func from(_ note: NoteRecord, reason: String) -> NoteVersionRecord {
        let now = ISO8601DateFormatter.jottery.string(from: Date())
        return NoteVersionRecord(
            versionKey: "\(note.id):\(note.version)",
            noteId: note.id,
            version: note.version,
            createdAt: note.modifiedAt,
            syncedAt: now,
            content: note.content,
            tags: note.tags,
            attachments: note.attachments,
            syntaxLanguage: note.syntaxLanguage,
            wordWrap: note.wordWrap,
            showPreview: note.showPreview,
            color: note.color,
            reason: reason,
            contentHash: note.contentHash,
            parentHash: note.parentHash
        )
    }
}

File: android-native/app/src/main/java/com/jottery/android/model/NoteVersionRecord.kt
package com.jottery.android.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import com.jottery.android.util.DateUtils

/**
 * Version snapshot of a note. Primary key is composite "noteId:version".
 */
@Entity(
    tableName = "note_versions",
    indices = [
        Index(value = ["note_id"]),
        Index(value = ["note_id", "version"]),
    ]
)
data class NoteVersionRecord(
    @PrimaryKey
    @ColumnInfo(name = "version_key")
    val versionKey: String,

    @ColumnInfo(name = "note_id")
    val noteId: String,

    val version: Int,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "synced_at")
    val syncedAt: String,

    // Encrypted JSON
    val content: String,

    // Encrypted JSON
    val tags: String,

    // JSON array
    val attachments: String,

    @ColumnInfo(name = "syntax_language")
    val syntaxLanguage: String? = null,

    @ColumnInfo(name = "word_wrap")
    val wordWrap: Boolean? = null,

    @ColumnInfo(name = "show_preview")
    val showPreview: Boolean? = null,

    val color: String? = null,

    val reason: String,

    @ColumnInfo(name = "content_hash")
    val contentHash: String? = null,

    @ColumnInfo(name = "parent_hash")
    val parentHash: String? = null,
) {
    companion object {
        fun from(note: NoteRecord, reason: String): NoteVersionRecord {
            val key = "${note.id}:${note.version}"
            return NoteVersionRecord(
                versionKey = key,
                noteId = note.id,
                version = note.version,
                createdAt = note.modifiedAt,
                syncedAt = DateUtils.nowISO8601(),
                content = note.content,
                tags = note.tags,
                attachments = note.attachments,
                syntaxLanguage = note.syntaxLanguage,
                wordWrap = note.wordWrap,
                showPreview = note.showPreview,
                color = note.color,
                reason = reason,
                contentHash = note.contentHash,
                parentHash = note.parentHash,
            )
        }
    }
}
